File: Toy-Compiler/python/zap/codegen.py
"""
Code generator: AST → flat bytecode for the Zap stack VM.

Jump offsets are signed integers relative to the instruction after the jump.
  JMP  n  →  ip += n          (unconditional)
  JF   n  →  if not pop(): ip += n
"""

from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Union
from .ast_nodes import *


class Op(Enum):
    PUSH_INT  = auto()
    PUSH_BOOL = auto()
    LOAD      = auto()
    STORE     = auto()
    ADD       = auto()
    SUB       = auto()
    MUL       = auto()
    DIV       = auto()
    EQ        = auto()
    NEQ       = auto()
    LT        = auto()
    GT        = auto()
    LEQ       = auto()
    GEQ       = auto()
    AND       = auto()
    OR        = auto()
    NOT       = auto()
    NEG       = auto()
    JMP       = auto()
    JMP_FALSE = auto()
    CALL      = auto()
    RET       = auto()
    PRINT     = auto()
    POP       = auto()
    HALT      = auto()


_BIN_OPS: dict[str, Op] = {
    "+": Op.ADD, "-": Op.SUB, "*": Op.MUL, "/": Op.DIV,
    "==": Op.EQ, "!=": Op.NEQ,
    "<":  Op.LT, ">":  Op.GT, "<=": Op.LEQ, ">=": Op.GEQ,
    "&&": Op.AND, "||": Op.OR,
}


@dataclass
class Instr:
    op:   Op
    arg:  Union[int, bool, str, None] = None
    arg2: Union[int, None] = None   # CALL only: arg=fn_name, arg2=nargs

    def __repr__(self) -> str:
        if self.arg2 is not None:
            return f"{self.op.name:<12} {self.arg} {self.arg2}"
        if self.arg is not None:
            return f"{self.op.name:<12} {self.arg!r}"
        return self.op.name


@dataclass
class FnChunk:
    params: list[str]
    code:   list[Instr] = field(default_factory=list)


@dataclass
class CompiledProgram:
    functions: dict[str, FnChunk] = field(default_factory=dict)
    entry:     str                = "__main__"

    def disassemble(self) -> str:
        lines: list[str] = []
        for name, chunk in self.functions.items():
            lines.append(f"fn {name}({', '.join(chunk.params)}):")
            for i, ins in enumerate(chunk.code):
                lines.append(f"  {i:4d}  {ins}")
        return "\n".join(lines)


class CodeGen:
    def __init__(self) -> None:
        self._prog  = CompiledProgram()
        self._chunk: FnChunk | None = None

    def compile(self, program: Program) -> CompiledProgram:
        # Collect all function decls first (forward references).
        for decl in program.decls:
            if isinstance(decl, FnDecl):
                self._start_fn(decl.name, [n for n, _ in decl.params])

        for decl in program.decls:
            if isinstance(decl, FnDecl):
                self._compile_fn(decl)
            else:
                self._ensure_main()
                self._compile_stmt(decl)

        self._ensure_main()
        main = self._prog.functions["__main__"]
        if not main.code or main.code[-1].op != Op.HALT:
            main.code.append(Instr(Op.HALT))

        # Determine entry: prefer user-defined 'main', else '__main__'
        if "main" in self._prog.functions:
            self._prog.entry = "main"
        return self._prog

    # ── internals ─────────────────────────────────────────────────────────────

    def _ensure_main(self) -> None:
        if "__main__" not in self._prog.functions:
            self._prog.functions["__main__"] = FnChunk(params=[])
        self._chunk = self._prog.functions["__main__"]

    def _start_fn(self, name: str, params: list[str]) -> None:
        self._prog.functions[name] = FnChunk(params=params)

    def _compile_fn(self, fn: FnDecl) -> None:
        prev = self._chunk
        self._chunk = self._prog.functions[fn.name]
        for stmt in fn.body.stmts:
            self._compile_stmt(stmt)
        code = self._chunk.code
        if not code or code[-1].op != Op.RET:
            code.append(Instr(Op.PUSH_BOOL, False))
            code.append(Instr(Op.RET))
        self._chunk = prev

    def _emit(self, op: Op, arg=None, arg2=None) -> None:
        assert self._chunk is not None
        self._chunk.code.append(Instr(op, arg, arg2))

    def _here(self) -> int:
        return len(self._chunk.code)

    def _patch(self, idx: int, target: int) -> None:
        # offset is from the instruction after the jump to target
        self._chunk.code[idx].arg = target - idx - 1

    # ── statements ────────────────────────────────────────────────────────────

    def _compile_stmt(self, stmt: Stmt) -> None:
        if isinstance(stmt, LetStmt):
            self._compile_expr(stmt.value)
            self._emit(Op.STORE, stmt.name)

        elif isinstance(stmt, ReturnStmt):
            if stmt.value is not None:
                self._compile_expr(stmt.value)
            else:
                self._emit(Op.PUSH_BOOL, False)
            self._emit(Op.RET)

        elif isinstance(stmt, IfStmt):
            self._compile_expr(stmt.cond)
            jf = self._here()
            self._emit(Op.JMP_FALSE, 0)           # placeholder
            for s in stmt.then_block.stmts:
                self._compile_stmt(s)
            if stmt.else_block:
                jmp = self._here()
                self._emit(Op.JMP, 0)             # skip else block
                self._patch(jf, self._here())
                for s in stmt.else_block.stmts:
                    self._compile_stmt(s)
                self._patch(jmp, self._here())
            else:
                self._patch(jf, self._here())

        elif isinstance(stmt, WhileStmt):
            loop = self._here()
            self._compile_expr(stmt.cond)
            jf = self._here()
            self._emit(Op.JMP_FALSE, 0)
            for s in stmt.body.stmts:
                self._compile_stmt(s)
            back = self._here()
            self._emit(Op.JMP, loop - back - 1)   # back-edge (negative offset)
            self._patch(jf, self._here())

        elif isinstance(stmt, PrintStmt):
            self._compile_expr(stmt.value)
            self._emit(Op.PRINT)

        elif isinstance(stmt, ExprStmt):
            self._compile_expr(stmt.expr)
            if not isinstance(stmt.expr, AssignExpr):
                self._emit(Op.POP)

    # ── expressions ───────────────────────────────────────────────────────────

    def _compile_expr(self, expr: Expr) -> None:
        if isinstance(expr, IntLit):
            self._emit(Op.PUSH_INT, expr.value)
        elif isinstance(expr, BoolLit):
            self._emit(Op.PUSH_BOOL, expr.value)
        elif isinstance(expr, IdentExpr):
            self._emit(Op.LOAD, expr.name)
        elif isinstance(expr, AssignExpr):
            self._compile_expr(expr.value)
            self._emit(Op.STORE, expr.name)
        elif isinstance(expr, BinOp):
            self._compile_expr(expr.left)
            self._compile_expr(expr.right)
            self._emit(_BIN_OPS[expr.op])
        elif isinstance(expr, UnaryOp):
            self._compile_expr(expr.operand)
            self._emit(Op.NOT if expr.op == "!" else Op.NEG)
        elif isinstance(expr, CallExpr):
            for arg in expr.args:
                self._compile_expr(arg)
            self._emit(Op.CALL, expr.name, len(expr.args))


def compile_program(program: Program) -> CompiledProgram:
    return CodeGen().compile(program)

File: modules/compiler/rust/src/codegen.rs
//! Code generator: AST → flat bytecode for the Zap stack VM.
//!
//! Jump offsets are signed integers relative to the instruction after the jump.
//! JMP  n  →  ip += n  (signed)
//! JF   n  →  if !pop(): ip += n

use std::collections::HashMap;
use crate::ast::*;

#[derive(Debug, Clone, PartialEq)]
pub enum Op {
    PushInt, PushBool,
    Load, Store,
    Add, Sub, Mul, Div,
    Eq, Neq, Lt, Gt, Leq, Geq,
    And, Or, Not, Neg,
    Jmp, JmpFalse,
    Call,
    Ret,
    Print,
    Pop,
    Halt,
}

#[derive(Debug, Clone)]
pub struct Instr {
    pub op:   Op,
    pub ival: i64,        // integer or bool (0/1) payload
    pub sval: String,     // string payload (Load/Store/Call)
    pub arg2: usize,      // CALL: nargs
}

impl Instr {
    fn int(op: Op, n: i64)  -> Self { Instr { op, ival: n, sval: String::new(), arg2: 0 } }
    fn bool(b: bool)         -> Self { Instr { op: Op::PushBool, ival: b as i64, sval: String::new(), arg2: 0 } }
    fn name(op: Op, s: &str) -> Self { Instr { op, ival: 0, sval: s.to_string(), arg2: 0 } }
    fn call(fn_name: &str, nargs: usize) -> Self {
        Instr { op: Op::Call, ival: 0, sval: fn_name.to_string(), arg2: nargs }
    }
    fn simple(op: Op) -> Self { Instr { op, ival: 0, sval: String::new(), arg2: 0 } }

    pub fn display(&self) -> String {
        match self.op {
            Op::PushInt  => format!("PUSH_INT  {}", self.ival),
            Op::PushBool => format!("PUSH_BOOL {}", self.ival != 0),
            Op::Load     => format!("LOAD      {}", self.sval),
            Op::Store    => format!("STORE     {}", self.sval),
            Op::Jmp      => format!("JMP       {}", self.ival),
            Op::JmpFalse => format!("JMP_FALSE {}", self.ival),
            Op::Call     => format!("CALL      {} {}", self.sval, self.arg2),
            _            => format!("{}", match self.op {
                Op::Add  => "ADD",  Op::Sub  => "SUB", Op::Mul  => "MUL", Op::Div => "DIV",
                Op::Eq   => "EQ",   Op::Neq  => "NEQ", Op::Lt   => "LT",  Op::Gt  => "GT",
                Op::Leq  => "LEQ",  Op::Geq  => "GEQ", Op::And  => "AND", Op::Or  => "OR",
                Op::Not  => "NOT",  Op::Neg  => "NEG", Op::Ret  => "RET", Op::Pop => "POP",
                Op::Halt => "HALT", Op::Print => "PRINT",
                _ => "???",
            }),
        }
    }
}

#[derive(Debug, Clone)]
pub struct FnChunk {
    pub params: Vec<String>,
    pub code:   Vec<Instr>,
}

#[derive(Debug, Clone)]
pub struct CompiledProgram {
    pub functions: HashMap<String, FnChunk>,
    pub entry:     String,
}

impl CompiledProgram {
    pub fn disassemble(&self) -> String {
        let mut out = String::new();
        for (name, chunk) in &self.functions {
            out.push_str(&format!("fn {}({}):\n", name, chunk.params.join(", ")));
            for (i, ins) in chunk.code.iter().enumerate() {
                out.push_str(&format!("  {:4}  {}\n", i, ins.display()));
            }
        }
        out
    }
}

pub struct CodeGen {
    prog:    CompiledProgram,
    current: String,   // name of function being compiled
}

impl CodeGen {
    pub fn new() -> Self {
        let mut functions = HashMap::new();
        functions.insert("__main__".to_string(), FnChunk { params: vec![], code: vec![] });
        CodeGen {
            prog: CompiledProgram { functions, entry: "__main__".to_string() },
            current: "__main__".to_string(),
        }
    }

    pub fn compile(mut self, program: &Program) -> CompiledProgram {
        // Pre-declare all functions so CALL can reference them.
        for decl in &program.decls {
            if let Decl::Fn(fn_decl) = decl {
                let params = fn_decl.params.iter().map(|(n, _)| n.clone()).collect();
                self.prog.functions.insert(
                    fn_decl.name.clone(),
                    FnChunk { params, code: vec![] },
                );
            }
        }

        for decl in &program.decls {
            match decl {
                Decl::Fn(fn_decl) => self.compile_fn(fn_decl),
                Decl::Stmt(stmt)  => {
                    self.current = "__main__".to_string();
                    self.compile_stmt(stmt);
                }
            }
        }

        // Ensure __main__ ends with HALT.
        self.current = "__main__".to_string();
        let code = &self.prog.functions["__main__"].code;
        if code.is_empty() || code.last().map(|i| &i.op) != Some(&Op::Halt) {
            self.emit(Instr::simple(Op::Halt));
        }

        // Prefer user-defined 'main' as entry.
        if self.prog.functions.contains_key("main") {
            self.prog.entry = "main".to_string();
        }

        self.prog
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    fn emit(&mut self, ins: Instr) {
        self.prog.functions.get_mut(&self.current).unwrap().code.push(ins);
    }

    fn here(&self) -> usize {
        self.prog.functions[&self.current].code.len()
    }

    fn patch(&mut self, idx: usize, target: usize) {
        let offset = target as i64 - idx as i64 - 1;
        self.prog.functions.get_mut(&self.current).unwrap().code[idx].ival = offset;
    }

    // ── functions ─────────────────────────────────────────────────────────────

    fn compile_fn(&mut self, fn_decl: &FnDecl) {
        self.current = fn_decl.name.clone();
        for stmt in &fn_decl.body.stmts {
            self.compile_stmt(stmt);
        }
        let code = &self.prog.functions[&self.current].code;
        if code.is_empty() || code.last().map(|i| &i.op) != Some(&Op::Ret) {
            self.emit(Instr::bool(false));
            self.emit(Instr::simple(Op::Ret));
        }
    }

    // ── statements ────────────────────────────────────────────────────────────

    fn compile_stmt(&mut self, stmt: &Stmt) {
        match stmt {
            Stmt::Let(s) => {
                self.compile_expr(&s.value);
                self.emit(Instr::name(Op::Store, &s.name));
            }
            Stmt::Return(s) => {
                match &s.value {
                    Some(e) => self.compile_expr(e),
                    None    => self.emit(Instr::bool(false)),
                }
                self.emit(Instr::simple(Op::Ret));
            }
            Stmt::If(s) => {
                self.compile_expr(&s.cond);
                let jf = self.here();
                self.emit(Instr::int(Op::JmpFalse, 0));
                for stmt in &s.then_block.stmts { self.compile_stmt(stmt); }
                if let Some(else_b) = &s.else_block {
                    let jmp = self.here();
                    self.emit(Instr::int(Op::Jmp, 0));
                    self.patch(jf, self.here());
                    for stmt in &else_b.stmts { self.compile_stmt(stmt); }
                    self.patch(jmp, self.here());
                } else {
                    self.patch(jf, self.here());
                }
            }
            Stmt::While(s) => {
                let loop_start = self.here();
                self.compile_expr(&s.cond);
                let jf = self.here();
                self.emit(Instr::int(Op::JmpFalse, 0));
                for stmt in &s.body.stmts { self.compile_stmt(stmt); }
                let back = self.here();
                let offset = loop_start as i64 - back as i64 - 1;
                self.emit(Instr::int(Op::Jmp, offset));
                self.patch(jf, self.here());
            }
            Stmt::Print(s) => {
                self.compile_expr(&s.value);
                self.emit(Instr::simple(Op::Print));
            }
            Stmt::Expr(s) => {
                self.compile_expr(&s.expr);
                if !s.expr.is_assign() {
                    self.emit(Instr::simple(Op::Pop));
                }
            }
        }
    }

    // ── expressions ───────────────────────────────────────────────────────────

    fn compile_expr(&mut self, expr: &Expr) {
        match expr {
            Expr::IntLit(v, _)  => self.emit(Instr::int(Op::PushInt, *v)),
            Expr::BoolLit(b, _) => self.emit(Instr::bool(*b)),
            Expr::Ident(n, _)   => self.emit(Instr::name(Op::Load, n)),
            Expr::Assign { name, value, .. } => {
                self.compile_expr(value);
                self.emit(Instr::name(Op::Store, name));
            }
            Expr::BinOp { op, left, right, .. } => {
                self.compile_expr(left);
                self.compile_expr(right);
                let op_code = match op.as_str() {
                    "+"  => Op::Add, "-"  => Op::Sub, "*"  => Op::Mul, "/" => Op::Div,
                    "==" => Op::Eq,  "!=" => Op::Neq,
                    "<"  => Op::Lt,  ">"  => Op::Gt, "<=" => Op::Leq, ">=" => Op::Geq,
                    "&&" => Op::And, "||" => Op::Or,
                    _    => panic!("unknown op: {op}"),
                };
                self.emit(Instr::simple(op_code));
            }
            Expr::UnaryOp { op, operand, .. } => {
                self.compile_expr(operand);
                self.emit(Instr::simple(if op == "!" { Op::Not } else { Op::Neg }));
            }
            Expr::Call { name, args, .. } => {
                for arg in args { self.compile_expr(arg); }
                self.emit(Instr::call(name, args.len()));
            }
        }
    }
}

pub fn compile_program(program: &Program) -> CompiledProgram {
    CodeGen::new().compile(program)
}
